README.md
TK102 GPS server - DEPRECATED
================

This package is no longer being maintained.

---

Receive and parse GPS data from Xexun TK102 trackers.

The Xexun TK102 is a GPS device that can send coordinates over TCP to 
a server via GPRS. This Node.js script creates a TCP server that listens 
for GPRMC data, parse it and send the data to your post-process function. 
The parsed data is provided in a clean easy to use object, so you can 
easily store it in a database or push to a websocket server, etc.


- [Example](#example)
- [Prepare device](#prepare-device)
- [Installation](#installation)
- [Settings](#settings)
- [Events](#events)
- [Notes](#notes)


Example
-------

```js
var server = require ('tk102');

// start server
server.createServer ({
  port: 1337
});

// incoming data, i.e. update a map
server.on ('track', function (gps) {
  updateMap (gps.geo.latitude, gps.geo.longitude);
});
```


Prepare device
--------------

Assuming your simcard has enough SMS and data credits and the TK102 is 
configured for your provider's APN, simply send `adminip123456 IP PORT` 
where obviously `IP` is the server's IP, `PORT` is the port to listen on 
and `123456` is your admin password. :) It cannot take hostnames as it 
has no dns features on board.

Activate sending coordinates: **t030s003n123456**

This tells the device to send its location AFTER each **30** seconds and 
no more than **3** times. 30 seconds is the minimum. 
Send `t030s***n123456` to go on for infinity.

* **s** (seconds) can also be **m** (minutes) or **h** (hours)
* To end tracking send `notn123456`


Installation
------------

Stable: `npm install tk102`

Develop: `npm install fvdm/nodejs-tk102#develop`


Settings
--------

name        | type    | default | description
:-----------|:--------|:--------|:----------------------------------
ip          | string  | 0.0.0.0 | Listen on IP, `0.0.0.0` is all IPs
port        | integer | 0       | Listen on port, `0` is random (see [listening](#listening) event)
connections | integer | 10      | Maximum simultaneous connections
timeout     | integer | 10      | Idle time out in seconds

```js
server.createServer ({
  ip: '1.2.3.4',
  port: 0
});
```


Events
------

The server emits the following events about the server status and incoming GPS pushes.


### track
**( gpsObject )**

The data push from the device.

```js
server.on ('track', function (gps) {
  { raw: '1203301642,0031698765432,GPRMC,144219.000,A,5213.0327,N,00516.7759,E,0.63,179.59,300312,,,A*6D,F,imei:123456789012345,123',
    datetime: '2012-03-30 16:42',
    phone: '0031698765432',
    gps: { date: '2012-03-30', time: '14:42:19.000', signal: 'full', fix: 'active' },
    geo: { latitude: 52.130326, longitude: 5.167759, bearing: 179 },
    speed: { knots: 0.63, kmh: 1.167, mph: 0.725 },
    imei: '123456789012345' }
})
```

property                 | description
:------------------------|:--------------------------------------------------
raw                      | the input string without trailing whitespace
datetime                 | the device 24h clock
phone                    | the admin phonenumber that initiated this tracking
imei                     | device IMEI
[gps](#gpsobjectgps)     | information about the GPS signal
[geo](#gpsobjectgeo)     | geographical position and direction
[speed](#gpsobjectspeed) | travel speed


#### gpsObject.gps

property    | description
:-----------|:---------------------------------------
date        | date as received from GPS
time        | time in 24h UTC as received from GPS
signal      | signal strength, either _full_ or _low_
fix         | GPS fix, either _active_ or _invalid_


#### gpsObject.geo

property    | description
:-----------|:--------------------
latitude    | position latitude
longitude   | position longitude
bearing     | direction in degrees


#### gpsObject.speed

property    | description
:-----------|:----------------------------------
knots       | speed in knots per hour (original)
kmh         | speed in kilometer per hour
mph         | speed in miles per hour


### data
**( rawString )**

The raw unprocessed inbound data.

```js
server.on ('data', function (raw) {
  console.log ('Incoming data: '+ raw);
});
```


### listening
**( listeningObject )**

Very useful to find out random port (0).

```js
server.on ('listening', function (listen) {
  // listen = { port: 56751, family: 2, address: '0.0.0.0' }
});
```


### connection
**( socket )**

Emitted when a connection is established with the server, includes the socket basics.

```js
server.on ('connection', function (socket) {
  console.log ('Connection from '+ socket.remoteAddress);
});
```


### disconnect
**( socket )**

Emitted when a connection is ended, includes the socket basics.

```js
server.on ('disconnect', function (socket) {
  console.log ('Disconnected device '+ socket.remoteAddress);
});
```


### timeout
**( socket )**

Emitted when a connection expires, includes the socket basics.

```js
server.on ('timeout', function (socket) {
  console.log ('Time-out from '+ socket.remoteAddress);
});
```


### fail
**( Error )**

Emitted when data cannot be parsed.
Useful for debugging device issues.

`Error` is an `instanceof Error` with .stack trace.

```js
server.on ('fail', function (err) {
  console.log (err);
});
```


### error
**( Error )**

Emitted when a server related error occured.

`Error` is an `instanceof Error` with .stack trace.


#### Messages

error                    | description
:------------------------|:---------------------------------
Server error             | Catch server failures
Socket error             | Catch communication failures
IP or port not available | This catches EADDRNOTAVAIL errors


```js
server.on ('error', function (err) {
  console.log (err);
});
```


### log
**( name, data )**

Emitted on any of the above events.
Useful for debugging and logging.

argument | type   | description
:--------|:-------|:------------------------
name     | string | Event name, i.e. `track`
value    | mixed  | Data from the event

```js
server.on ('log', function (name, value) {
  console.log ('Event: ' + name);
  console.log (value);
});
```


Notes
-----

There is no security built in, anyone could push data to your server.

The parsed object is based on a lot of testing in the field and
incomplete documentation, I hope I got it right. There are many
variations of the TK102 tracker available with each different data
formats. It may be possible your device or even mine is one of the
clones and incompatible with this module.

If you can, please provide multiple responses from the [data](#data)
event in a [Github issue](https://github.com/fvdm/nodejs-tk102/issues)
(public) or contact me [directly](https://frankl.in/contact), your data
is handled confidentially. Preferably leave the device tracking with an
interval of 30 seconds or less for about 3 to 5 minutes outdoors or near
a window. This way I can figure out the data format and various states
the tracker reports about itself to the server.


Unlicense
---------

This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <https://unlicense.org>


Author
------

[Franklin](https://fvdm.com)
| [Buy me a coffee](https://fvdm.com/donating)


CHANGELOG.md
#### 1.3.4 (2021-07-17)

#### 1.3.3 (2021-07-17)

#### 1.3.2 (2021-07-17)

##### Chores

* **package:**
  *  Update dotest dev dep ([50de1e95](https://github.com/fvdm/nodejs-tk102/commit/50de1e958b194756e2dcb2a8ab8862ef03b507ac))
  *  Update dev deps ([e590be47](https://github.com/fvdm/nodejs-tk102/commit/e590be47c702824524122ce6d25fa92dbc5a2f63))
  *  Replaced test runner and dev deps by dotest ([37bbf7a0](https://github.com/fvdm/nodejs-tk102/commit/37bbf7a0f61d70a04a1da579377c0865c07e3a91))
* **develop:**
  *  Added gitignore config ([e352157f](https://github.com/fvdm/nodejs-tk102/commit/e352157f3e8bf1f0c1dfe5cb29f293a82bb4814b))
  *  Added bitHound config ([c702663b](https://github.com/fvdm/nodejs-tk102/commit/c702663bbe7cf7c01a7135974835d440b4df4a05))

##### Documentation Changes

* **badges:**
  *  Removed broken npm badge ([c54d9d0f](https://github.com/fvdm/nodejs-tk102/commit/c54d9d0f4cdf4a6553db5376cc4a692b8ccba81c))
  *  Add Gemnasium dependencies status ([e3d1ef9b](https://github.com/fvdm/nodejs-tk102/commit/e3d1ef9b56ee8f344169f81b4211f208c3986fe1))
  *  Add npm version for changelog ([fba6684e](https://github.com/fvdm/nodejs-tk102/commit/fba6684e7dc95e59861544b08960d48b7c42ade8))
*  Deprecation message ([4cc2bd95](https://github.com/fvdm/nodejs-tk102/commit/4cc2bd95524d7b71b4970af86c22a94843dd52b2))
* **readme:**
  *  Added links to gpsObject subs ([f6c15076](https://github.com/fvdm/nodejs-tk102/commit/f6c150767bf922038af767d53b410ceba8784a62))
  *  Fix legacy tables ([71086334](https://github.com/fvdm/nodejs-tk102/commit/71086334a8bd20484113463c15ee3937edb3a9ee))

##### Other Changes

*  always run both test commands ([1235f254](https://github.com/fvdm/nodejs-tk102/commit/1235f254af0dacbaf0b93ef479888e58e0684cfb))
*  added Tonic example code ([400cc9aa](https://github.com/fvdm/nodejs-tk102/commit/400cc9aa30ea886da9fe11139e69b44086dc121c))
*  added eslint to npm test ([a139e59c](https://github.com/fvdm/nodejs-tk102/commit/a139e59c76812f62763f5f8270069bde21e656fe))
*  added dotest dev dependency ([e2b7d3c8](https://github.com/fvdm/nodejs-tk102/commit/e2b7d3c800b41e2496ded9de2b0f6e028b8a2df2))
*  added contributors ([605c0c5f](https://github.com/fvdm/nodejs-tk102/commit/605c0c5fcbb3ae085b32d6f8fee4d712b785dac7))
*  fixed text typo ([3f383e3f](https://github.com/fvdm/nodejs-tk102/commit/3f383e3f6d1df207c8a5fcea3e7f95b042a4f15c))

##### Refactors

* **package:**  Minimum supported node v4.0 ([19725445](https://github.com/fvdm/nodejs-tk102/commit/197254450b513b8c20a888cfaa156604735ab243))

##### Code Style Changes

* **comments:**
  *  Minor tweaks ([bdcbc9f5](https://github.com/fvdm/nodejs-tk102/commit/bdcbc9f507e6618754dcc0b01d2a4620e67d1b5f))
  *  Add JSDoc to functions ([80529698](https://github.com/fvdm/nodejs-tk102/commit/805296982910624a48c908d5d1065fca4b466396))

##### Tests

* **package:**  Change dev deps to dotest ([e3eab190](https://github.com/fvdm/nodejs-tk102/commit/e3eab19079d3a45386dce01b588c074eeeb0dab2))
* **config:**
  *  Update Travis CI node versions ([88613cc6](https://github.com/fvdm/nodejs-tk102/commit/88613cc60dd130481cc4b1ff1f846745be713633))
  *  Travis CI update node versions ([affe971b](https://github.com/fvdm/nodejs-tk102/commit/affe971b85b1ab48718254a5dab972a789310c3d))
  *  Use dynamic node versions on Travis CI ([336660b0](https://github.com/fvdm/nodejs-tk102/commit/336660b0d40fa0befaa6dec267379c69ac8a1b94))
* **lint:**  Update eslint to ES6 ([2813b1c7](https://github.com/fvdm/nodejs-tk102/commit/2813b1c76072fb3b2078eb938eaaa5d43df3972c))
*  add node v6 to Travis config ([d5870eb5](https://github.com/fvdm/nodejs-tk102/commit/d5870eb55bf0a1b6169ae81371c13058a1d73203))
*  fixed syntax typo ([cd7ea876](https://github.com/fvdm/nodejs-tk102/commit/cd7ea876fb758fd7980016c167916f28820b9ff8))
*  added basic .closeServer method check ([2b8a10ac](https://github.com/fvdm/nodejs-tk102/commit/2b8a10ac01fd092f1e075b264f3511020d129fea))
*  removed redundant checks ([951b19ee](https://github.com/fvdm/nodejs-tk102/commit/951b19ee6457f38f2aea9626184f201a69fc1c49))
*  updated eslint config ([1c4afcee](https://github.com/fvdm/nodejs-tk102/commit/1c4afcee15fde1a3602d30656a0477999a68826c))
*  fixed horrible typos ([5e265705](https://github.com/fvdm/nodejs-tk102/commit/5e265705dfac524d017eef0c459d5358be188431))
*  removed old node from Travis config ([63c43563](https://github.com/fvdm/nodejs-tk102/commit/63c4356385092060fe405be5603e5090171ef757))
*  replaced script with dotest module ([7620fb92](https://github.com/fvdm/nodejs-tk102/commit/7620fb9278aa3cdc2685b65696cafdd5aedd559a))



test.js
/*
Name:         tk102 - test.js
Description:  Test script for TK102 GPS server for node.js
Source:       https://github.com/fvdm/nodejs-tk102
License:      Unlicense / Public Domain (see UNLICENSE file)
*/

var dotest = require ('dotest');
var app = require ('./');

// Str to work with
var input = '1203292316,0031698765432,GPRMC,211657.000,A,5213.0247,N,00516.7757,E,0.00,273.30,290312,,,A*62,F,imei:123456789012345,123';


// module
dotest.add ('Module', function () {
  dotest.test ()
    .isObject ('fail', 'exports', app)
    .isObject ('fail', '.settings', app && app.settings)
    .isFunction ('fail', '.event', app && app.event)
    .isFunction ('fail', '.createServer', app && app.createServer)
    .isFunction ('fail', '.closeServer', app && app.closeServer)
    .isFunction ('fail', '.fixGeo', app && app.fixGeo)
    .isFunction ('fail', '.checksum', app && app.checksum)
    .isFunction ('fail', '.parse', app && app.parse)
    .done ();
});

// checksum valid
dotest.add ('checksum valid', function () {
  var data = app.checksum (input);

  dotest.test ()
    .isExactly ('fail', 'data', data, true)
    .done ();
});

// checksum invalid
dotest.add ('checksum invalid', function () {
  var data = app.checksum (input.toLowerCase ());

  dotest.test ()
    .isExactly ('fail', 'data', data, false)
    .done ();
});

// parser valid
dotest.add ('parse valid', function () {
  var data = app.parse (input);

  dotest.test ()
    .isObject ('fail', 'data', data)
    .isExactly ('fail', 'data.raw', data && data.raw, input)
    .isExactly ('fail', 'data.checksum', data && data.checksum, true)
    .isExactly ('fail', 'data.phone', data && data.phone, '0031698765432')
    .isExactly ('fail', 'data.imei', data && data.imei, '123456789012345')
    .isExactly ('fail', 'data.datetime', data && data.datetime, '2012-03-29 23:16')
    .isObject ('fail', 'data.gps', data && data.gps)
    .isExactly ('fail', 'data.gps.date', data && data.gps && data.gps.date, '2012-03-29')
    .isExactly ('fail', 'data.gps.time', data && data.gps && data.gps.time, '21:16:57.000')
    .isExactly ('fail', 'data.gps.signal', data && data.gps && data.gps.signal, 'full')
    .isExactly ('fail', 'data.gps.fix', data && data.gps && data.gps.fix, 'active')
    .isObject ('fail', 'data.geo', data && data.geo)
    .isExactly ('fail', 'data.geo.latitude', data && data.geo && data.geo.latitude, 52.217078)
    .isExactly ('fail', 'data.geo.longitude', data && data.geo && data.geo.longitude, 5.279595)
    .isExactly ('fail', 'data.geo.bearing', data && data.geo && data.geo.bearing, 273)
    .isObject ('fail', 'data.speed', data && data.speed)
    .isExactly ('fail', 'data.speed.knots', data && data.speed && data.speed.knots, 0)
    .done ();
});

// parser fail
dotest.add ('parse fail', function () {
  var data = app.parse ('invalid input');

  dotest.test ()
    .isNull ('fail', 'data', data)
    .done ();
});


// Start the tests
dotest.run ();


example.js
/*
  Start with:  node example.js
  Telnet to:   telnet 127.0.0.1 1337
  Copy/paste:  1203292316,0031698765432,GPRMC,211657.000,A,5213.0247,N,00516.7757,E,0.00,273.30,290312,,,A*62,F,imei:123456789012345,123
*/

var tk102 = require ('tk102');
var net = require ('net');

var gps = '1203292316,0031698765432,GPRMC,211657.000,A,5213.0247,N,00516.7757,E,0.00,273.30,290312,,,A*62,F,imei:123456789012345,123';

// fancy console log
function output (data) {
  console.log ('\nIncoming GPS data:\n');
  console.dir (data, {
    colors: String (process.env.TERM) .match (/color$/)
  });
}

// report only track event to console
tk102.on ('track', output);

// wait for server to be ready
tk102.on ('listening', function (lst) {
  var client;

  console.log ('TK102 server is ready');

  // Send data with telnet
  client = net.connect (lst.port, function () {
    console.log ('Connected to TK102 server');
    console.log ('Sending GPS data string for processing');

    client.write (gps + '\r\n');
    client.end ();

    console.log ('CTRL+C to exit');
  });
});

// start server
tk102.createServer ({
  port: 1337
});


tk102.js
/*
Name:         tk102
Description:  TK102 GPS server for Node.js
Author:       Franklin (https://fvdm.com)
Source:       https://github.com/fvdm/nodejs-tk102
License:      Unlicense (Public Domain, see UNLICENSE file)
*/

var net = require ('net');
var EventEmitter = require ('events') .EventEmitter;
var tk102 = new EventEmitter ();

// device data
var specs = [
  function (raw) {
    // 1203292316,0031698765432,GPRMC,211657.000,A,5213.0247,N,00516.7757,E,0.00,273.30,290312,,,A*62,F,imei:123456789012345,123
    var result = null;
    var str = [];
    var datetime = '';
    var gpsdate = '';
    var gpstime = '';

    try {
      raw = raw.trim ();
      str = raw.split (',');

      if (str.length === 18 && str [2] === 'GPRMC') {
        datetime = str [0] .replace (/([0-9]{2})([0-9]{2})([0-9]{2})([0-9]{2})([0-9]{2})/, function (s, y, m, d, h, i) {
          return '20' + y + '-' + m + '-' + d + ' ' + h + ':' + i;
        });

        gpsdate = str [11] .replace (/([0-9]{2})([0-9]{2})([0-9]{2})/, function (s, d, m, y) {
          return '20' + y + '-' + m + '-' + d;
        });

        gpstime = str [3] .replace (/([0-9]{2})([0-9]{2})([0-9]{2})\.([0-9]{3})/, function (s0, h, i, s, ms) {
          return h + ':' + i + ':' + s + '.' + ms;
        });

        result = {
          raw: raw,
          datetime: datetime,
          phone: str [1],
          gps: {
            date: gpsdate,
            time: gpstime,
            signal: str [15] === 'F' ? 'full' : 'low',
            fix: str [4] === 'A' ? 'active' : 'invalid'
          },
          geo: {
            latitude: tk102.fixGeo (str [5], str [6]),
            longitude: tk102.fixGeo (str [7], str [8]),
            bearing: parseInt (str [10], 10)
          },
          speed: {
            knots: Math.round (str [9] * 1000) / 1000,
            kmh: Math.round (str [9] * 1.852 * 1000) / 1000,
            mph: Math.round (str [9] * 1.151 * 1000) / 1000
          },
          imei: str [16] .replace ('imei:', ''),
          checksum: tk102.checksum (raw)
        };
      }
    } catch (e) {
      result = null;
    }

    return result;
  }
];


// defaults
tk102.settings = {
  ip: '0.0.0.0',
  port: 0,
  connections: 10,
  timeout: 10
};


/**
 * Emit an event
 * and duplicate to 'log' event
 *
 * @param   {string}  name   Event name
 * @param   {string}  value  Event value
 * @return  {void}
 */

tk102.event = function (name, value) {
  tk102.emit (name, value);
  tk102.emit ('log', name, value);
};


/**
 * Catch uncaught exceptions (server kill)
 *
 * @param   {Error}  err  Error cause
 * @return  {void}
 */

process.on ('uncaughtException', function (err) {
  var error = new Error ('uncaught exception');

  error.error = err;
  console.log (error);
  tk102.event ('error', error);
});


/**
 * Create server
 *
 * @param   {object}  [vars]                 Override default settings
 * @param   {string}  [vars.ip='0.0.0.0']    Listen on IP
 * @param   {number}  [vars.port=0]          Listen on port, `0` = random
 * @param   {number}  [vars.connections=10]  Max server connections
 * @param   {number}  [vars.timeout=10]      Socket timeout in seconds
 * @return  {object}  tk102                  The server
 */

tk102.createServer = function (vars) {
  var key;

  // override settings
  if (typeof vars === 'object' && Object.keys (vars) .length >= 1) {
    for (key in vars) {
      tk102.settings [key] = vars [key];
    }
  }

  // start server
  tk102.server = net.createServer ();

  // maximum number of slots
  tk102.server.maxConnections = tk102.settings.connections;

  // server started
  tk102.server.on ('listening', function () {
    tk102.event ('listening', tk102.server.address ());
  });

  // inbound connection
  tk102.server.on ('connection', function (socket) {
    var connection = tk102.server.address ();

    connection.remoteAddress = socket.remoteAddress;
    connection.remotePort = socket.remotePort;

    tk102.event ('connection', connection);
    socket.setEncoding ('utf8');

    if (tk102.settings.timeout > 0) {
      socket.setTimeout (parseInt (tk102.settings.timeout * 1000, 10));
    }

    socket.on ('timeout', function () {
      tk102.event ('timeout', connection);
      socket.destroy ();
    });

    socket.on ('data', function (data) {
      var gps = {};
      var err = null;

      data = data.trim ();
      tk102.event ('data', data);

      if (data !== '') {
        gps = tk102.parse (data);

        if (gps) {
          tk102.event ('track', gps);
        } else {
          err = new Error ('Cannot parse GPS data from device');
          err.reason = err.message;
          err.input = data;
          err.connection = connection;

          tk102.event ('fail', err);
        }
      }
    });

    socket.on ('close', function (hadError) {
      connection.hadError = hadError;
      tk102.event ('disconnect', connection);
    });

    // error
    socket.on ('error', function (error) {
      var err = new Error ('Socket error');

      err.reason = error.message;
      err.socket = socket;
      err.settings = tk102.settings;

      tk102.event ('error', err);
    });
  });

  tk102.server.on ('error', function (error) {
    var err = new Error ('Server error');

    if (error === 'EADDRNOTAVAIL') {
      err = new Error ('IP or port not available');
    }

    err.reason = error.message;
    err.input = tk102.settings;

    tk102.event ('error', err);
  });

  // Start listening
  tk102.server.listen (tk102.settings.port, tk102.settings.ip);

  return tk102;
};


/**
 * Graceful close server
 *
 * @callback  callback
 * @param     {function}  callback  `(err)`
 * @return    {void}
 */

tk102.closeServer = function (callback) {
  if (!tk102.server) {
    callback (new Error ('server not started'));
    return;
  }

  tk102.server.close (callback);
};


/**
 * Parse GPRMC string
 *
 * @param   {string}  raw   Command text from device
 * @return  {object}  data  Parsed data
 */

tk102.parse = function (raw) {
  var data = null;
  var i = 0;

  while (data === null && i < specs.length) {
    data = specs [i] (raw);
    i++;
  }

  return data;
};


/**
 * Clean geo positions, with 6 decimals
 *
 * @param   {string}  one  Geo position
 * @param   {string}  two  Geo direction: N W S E
 * @return  {float}        Decimal geo position
 */

tk102.fixGeo = function (one, two) {
  var minutes = one.substr (-7, 7);
  var degrees = parseInt (one.replace (minutes, ''), 10);

  one = degrees + (minutes / 60);
  one = parseFloat ((two === 'S' || two === 'W' ? '-' : '') + one);

  return Math.round (one * 1000000) / 1000000;
};


/**
 * Check checksum in raw string
 *
 * @param   {string}   raw    Command text from device
 * @return  {boolean}  check  Checksum result
 */

tk102.checksum = function (raw) {
  var str = raw.trim () .split (/[,*#]/);
  var strsum = parseInt (str [15], 10);
  var strchk = str.slice (2, 15) .join (',');
  var check = 0;
  var i;

  for (i = 0; i < strchk.length; i++) {
    check ^= strchk.charCodeAt (i);
  }

  check = parseInt (check.toString (16), 10);
  return (check === strsum);
};


// ready
module.exports = tk102;
